File: problems/group_anagrams/group_anagrams_test.py
"""Test cases for the group_anagrams function."""

import pytest

from group_anagrams import groupAnagrams


@pytest.mark.parametrize(
    "strs, expected",
    [
        (
            ["eat", "tea", "tan", "ate", "nat", "bat"],
            [["bat"], ["tan", "nat"], ["eat", "tea", "ate"]],
        ),  # Basic case
        ([""], [[""]]),  # Single empty string
        (["a"], [["a"]]),  # Single character
        (
            ["abc", "bca", "cab", "xyz", "zyx", "yxz"],
            [["abc", "bca", "cab"], ["xyz", "zyx", "yxz"]],
        ),  # Multiple anagram groups
        (
            ["listen", "silent", "enlist", "inlets", "google", "gogole"],
            [["listen", "silent", "enlist", "inlets"], ["google", "gogole"]],
        ),  # Longer strings
        (
            ["aabb", "abab", "bbaa", "baba", "abba"],
            [["aabb", "abab", "bbaa", "baba", "abba"]],
        ),  # All are anagrams
        (
            ["rat", "tar", "art", "star", "tars"],
            [["rat", "tar", "art"], ["star", "tars"]],
        ),  # Mixed lengths
        ([], []),  # Empty list
    ],
    ids=[
        "basic_case",
        "single_empty_string",
        "single_character",
        "multiple_anagram_groups",
        "longer_strings",
        "all_anagrams",
        "mixed_lengths",
        "empty_list",
    ],
)
def test_group_anagrams(strs, expected):
    result = groupAnagrams(strs)
    # Sort inner lists and the outer list for comparison
    result = [sorted(group) for group in result]
    result.sort()
    expected = [sorted(group) for group in expected]
    expected.sort()
    assert result == expected

File: problems/group_anagrams/group_anagrams_test.cc
#include "group_anagrams.h"

#include <gtest/gtest.h>
#include <algorithm>
#include <ranges>  // NOLINT(misc-include-cleaner) needed for std::ranges::sort
#include <string>
#include <vector>

struct GroupAnagramsCase {
    std::string test_name;
    std::vector<std::string> strs;
    std::vector<std::vector<std::string>> expected;
};

using GroupAnagramsTest = ::testing::TestWithParam<GroupAnagramsCase>;

namespace {
// Helper function to sort and compare results
bool compareResults(std::vector<std::vector<std::string>> result,
                    std::vector<std::vector<std::string>> expected) {
    // Sort inner vectors and outer vector for comparison
    for (auto& group : result) {
        std::ranges::sort(group);
    }
    std::ranges::sort(result);

    for (auto& group : expected) {
        std::ranges::sort(group);
    }
    std::ranges::sort(expected);

    return result == expected;
}
}  // namespace

TEST_P(GroupAnagramsTest, TestCases) {
    const GroupAnagramsCase& testCase = GetParam();
    std::vector<std::string> input = testCase.strs;  // Make a copy since function might modify
    const std::vector<std::vector<std::string>> result = groupAnagrams(input);
    EXPECT_TRUE(compareResults(result, testCase.expected));
}

INSTANTIATE_TEST_SUITE_P(
    GroupAnagramsTestCases, GroupAnagramsTest,
    ::testing::Values(
        GroupAnagramsCase{.test_name = "BasicCase",
                          .strs = {"eat", "tea", "tan", "ate", "nat", "bat"},
                          .expected = {{"bat"}, {"tan", "nat"}, {"eat", "tea", "ate"}}},
        GroupAnagramsCase{.test_name = "SingleEmptyString", .strs = {""}, .expected = {{""}}},
        GroupAnagramsCase{.test_name = "SingleCharacter", .strs = {"a"}, .expected = {{"a"}}},
        GroupAnagramsCase{.test_name = "MultipleAnagramGroups",
                          .strs = {"abc", "bca", "cab", "xyz", "zyx", "yxz"},
                          .expected = {{"abc", "bca", "cab"}, {"xyz", "zyx", "yxz"}}},
        GroupAnagramsCase{
            .test_name = "LongerStrings",
            .strs = {"listen", "silent", "enlist", "inlets", "google", "gogole"},
            .expected = {{"listen", "silent", "enlist", "inlets"}, {"google", "gogole"}}},
        GroupAnagramsCase{.test_name = "AllAnagrams",
                          .strs = {"aabb", "abab", "bbaa", "baba", "abba"},
                          .expected = {{"aabb", "abab", "bbaa", "baba", "abba"}}},
        GroupAnagramsCase{.test_name = "MixedLengths",
                          .strs = {"rat", "tar", "art", "star", "tars"},
                          .expected = {{"rat", "tar", "art"}, {"star", "tars"}}},
        GroupAnagramsCase{.test_name = "EmptyList", .strs = {}, .expected = {}}),
    [](const ::testing::TestParamInfo<GroupAnagramsCase>& info) { return info.param.test_name; });